### DISCUSSION TITLE
Modbus bridge : forward responses in passive listener / sniffer mode to MQTT

### DISCUSSION THREAD

**Discussion by relyd** [arendst/Tasmota#18618] [Ideas]
Referring to the excellent work of @jeroenst  at  https://tasmota.github.io/docs/Modbus-Bridge :

**Situation**:

- There is a (mandatory) Modbus Master (eg Solar Inverter, EV Charger etc) and a Slave ( eg. power meter) 
- We do not control the Master. Master continuously polls the slave (eg Voltage)
- Only one master is allowed on the bus

We would like to like to fetch the responses from the slave and publish for further processing/monitoring, i.e. 
- Listen for any response
- Publish response to MQTT

**Suggested :**

- Tasmota Modbus Bridge listens passively on Modbus for responses only (ModbusMonitor mode)
- You may start/stop Monitor Mode by sending a command like _ModbusMonitor2MqttStart_ , _ModbusMonitor2MqttStop_ 
- Starting ModbusMonitor mode would inhibit sending Modbus requests.

- Publish values **{[value1,value2,valueN]}** in received responses to a MQTT topic
- Suggested topic for values /ModbusBridge/ModbusReceived/DeviceAddress/FunctionCode/StartAddress/ eg 
_/ModbusBridge/ModbusReceived/1/5/1/_

- Publish raw (hex) responses to a MQTT topic
- Suggested topic for raw /ModbusBridge/ModbusReceivedRaw/

Please let me know your thoughts,

**Comment by jeroenst** [comment]
Maybe I can create that one day, your request is clear to me. 

Separate topics will not be available, messages will be send in json to the same topic (tasmota design). These can be manipulated with nodered for example. 

Data wil indeed always be raw bytes because the format is unknown. 

Command will be modbusmonitorstart and modbusmonitorstop imo.

Only drawback, I don't have a setup to test this at the moment.

**Comment by jeroenst** [comment]
1st try, not tested: https://github.com/jeroenst/Tasmota/tree/modbusmonitor

Responses are raw values (not hex).

**Reply by analystcmyk** [reply]
I am really sorry I missed the update, will compile and test....

**Comment by pepelillo** [comment]
I would like to test your code, can you give a clue on how to use it? I need a monitor mode, have to put a listener in the middle and it seems to be what I need.

**Reply by jeroenst** [reply]
You can compile https://github.com/jeroenst/Tasmota/tree/modbusmonitor and install the binairy.

Use modbusmonitorstart command in the console to start monitoring the modbus channel.

**Reply by jeroenst** [reply]
Do you have any progress on testing? Did you manage to compile the modbusmonitor branch or do you need a binairy for esp8266 or esp32 ?

**Comment by pepelillo** [comment]
Yes, i have a binary in a wemos,  is a 12.5 versión......is weird because the first time i launched the command i see some entries in my mqtt Explorer but after that any updates in the data.

I've tried the standard modbustcp Bridge and works as expected. 

My aim is to have the monitor working and suscribe to some data without using the estándar Bridge and make active queries. The reason is avoid the double máster issues in my bus (i have a máster already id0 and a slave of course id1)

I do not see any log in the console when It worked.

**Reply by jeroenst** [reply]
What do you see in the console of Tasmota, are there any modbus errors displayed?

**Reply by jeroenst** [reply]
Are you using modbusbridge and modbusmonitor simultaneously? 

That is not working (yet).

**Comment by pepelillo** [comment]
In monitor mode nothing at all. Even when I Saw data in mqtt Explorer.

**Reply by jeroenst** [reply]
What is the Loglevel in the console? You can try weblog 4

**Comment by pepelillo** [comment]
No. I started monitor mode first when It stops working i reset tasmota and uses Bridge mode. Bridge mode worked fine, but the multimaster issues I found.

**Comment by pepelillo** [comment]
I think tha this monitor mode would be very usefull when some kind of not máster Mode is needed.

**Comment by analystcmyk** [comment]
I compiled [the branch ](https://github.com/jeroenst/Tasmota/tree/modbusmonitor) with gitpod, for ESP-8266 , reporting 

`Tasmota Version 13.2.0.2(tasmota)-2_7_4_9(2023-11-16T09:28:14)`

`modbusmonitorstart` (loglevel 4) gives 

`
17:19:49.791 CMD: Grp 0, Cmd 'MODBUSMONITORSTART', Idx 1, Len 0, Pld -99, Data ''
17:19:49.796 RSL: RESULT = {"Command":"Unknown"}
`

My question in order to proceed testing : 

- Should this even run on 8266? Considering the modbusbridge should technically not be run on an ESP8266
- I have trouble setting the right compiler options/flags. Could someone add the compiled assets (bin)? 
- If possible I would run this on an ESP8266, if not ESP32.

Thank you,

**Reply by AndresL71** [reply]
did you configure your pins as ModBrTx Rx and Tx ENA?

**Reply by jeroenst** [reply]
Modbusbridge is currently in use on an esp8266 with a solar inverter at a friend of mine which works flawless. But for monitor mode I don't know if the functionality is ready for use because I don't have a test setup for it.

**Reply by ghost** [reply]
Can you explain how to config of the pins has to be done in detail?
Which one have to be used?
ModBr Tx
ModBr Tx ENA
ModBr Rx

**Comment by AndresL71** [comment]
Only getting CRC Errors
in monitor mode in between the master and slave (inverter--meter)
using a rs485 with gnd--DE-RE

16:47:58.216 MBS: MBR Driver receive error 9
16:47:58.573 MBS: MBR Driver receive error 9
16:47:59.134 MBS: MBR Driver receive error 9
16:47:59.475 MBS: MBR Driver receive error 9
16:47:59.633 MBS: MBR Driver receive error 9
16:47:59.778 MBS: MBR Driver receive error 9

any suggestions?

**Reply by jeroenst** [reply]
As I don't have a setup (functionality was on request) I don't have the option to debug this. But it seems that the crc is incorrect on the received packets.

**Reply by ghost** [reply]
same for me..

**Reply by FulvioSpelta** [reply]
Is it possible to see the hex dump of the modbus frames that have been received ?

**Reply by jeroenst** [reply]
You can define TASMOTAMODBUSDEBUG to see the RX hex values dumped to the console.

https://github.com/arendst/Tasmota/blob/9aa5a6d7fb8cfe60ec79b40269b1d27e906eb812/lib/lib_basic/TasmotaModbus-3.6.0/src/TasmotaModbus.cpp#L233

**Reply by ghost** [reply]
I tried to do that...

added the following im my user_config_override.h: 
#ifndef TASMOTAMODBUSDEBUG
#define TASMOTAMODBUSDEBUG
#endif

Compiled and flashed my esp32 again
After reboot I still see the CRC errors:

09:13:38.421 MBS: MBR Driver receive error 9
09:13:38.614 MBS: MBR Driver receive error 9
09:13:38.941 MBS: MBR Driver receive error 9
09:13:39.186 MBS: MBR Driver receive error 9
09:13:39.331 MBS: MBR Driver receive error 9

How can I see the Hex dump?

**Reply by jeroenst** [reply]
Did you issue the command "weblog 4" in order to show more debug messages?

**Reply by ghost** [reply]
09:38:32.258 MBS: MBR Driver receive error 9
09:38:32.403 MBS: MBR Driver receive error 9
09:38:32.595 MBS: MBR Driver receive error 9
09:38:32.708 CMD: weblog 4
09:38:32.710 SRC: WebConsole from 10.0.0.54
09:38:32.712 CMD: Grp 0, Cmd 'WEBLOG', Idx 1, Len 1, Pld 4, Data '4'
09:38:32.719 MQT: stat/meter/RESULT = {"WebLog":4}
09:38:32.923 MBS: MBR Driver receive error 9
09:38:33.167 MBS: MBR Driver receive error 9
09:38:33.311 MBS: MBR Driver receive error 9
09:38:33.503 MBS: MBR Driver receive error 9
09:38:33.775 MBS: MBR Driver receive error 9

**Reply by ghost** [reply]
do I miss any other components? I'm using your latest c commit: https://github.com/jeroenst/Tasmota/tree/modbusmonitor
Just added the TASMOTAMODBUSDEBUG as mentioned before

**Reply by jeroenst** [reply]
Try to override the define in tasmotamodbus.cpp

**Reply by ghost** [reply]
I tried to comment out the #ifdef TASMOTAMODBUSDEBUG and #endif lines, recompiled, but it hasn't changed anything.
Still the same behaviour.....
Just to be sure that everything is correct:
In "configure Module" I defined the 2 Pins (Pin 16 & 27) as ModBR RX and ModBR TX, commands in console: weblog 4 and modbusmonitorstart

**Reply by jeroenst** [reply]
I ordered an ESP32 from aliexpress, I will do testing and debugging when it arrives over 2 weeks

**Reply by danielkucera** [reply]
The issue is most likely in here:
https://github.com/arendst/Tasmota/blob/development/lib/lib_basic/TasmotaModbus-3.6.0/src/TasmotaModbus.cpp#L217
Tasmota only start reading the message once it sees expected device ID.

**Reply by danielkucera** [reply]
I have a similar problem:
Serial monitor shows:
```
02 03 00 02 08 02 62 38 - 02 03 00 02 08 00 E3 F9    ......b8  ........
```
but tasmota says:
```
16:53:47.992 MBS: Serial Received: 00 02 08 02 62 38 
16:53:47.993 MBS: MBR Driver receive error 9
16:53:50.349 MBS: Serial Received: 00 02 08 00 E3 F9 
16:53:50.349 MBS: MBR Driver receive error 9
```
this is most likely because mb_address = 0

**Reply by danielkucera** [reply]
@jeroenst , this works for me:
https://github.com/danielkucera/Tasmota/tree/modbus

How can we upstream the change in TasmotaModbus?

**Reply by AndresL71** [reply]
@jeroenst , this works for me: https://github.com/danielkucera/Tasmota/tree/modbus

Tried your repo today
still getting the same CRC error

19:33:32.781 MBS: Serial Received: 0B 03 00 06 10 00 02 0F 06 1F 0B 03 04 1E 0A 1E 1C 14 11 03
19:33:32.783 MBS: MBR Driver receive error 9
19:33:33.081 MBS: Serial Received: 0B 03 00 06 10 00 02 0F 06 1F 0B 03 04 1E 1A 1E 0B 11 00 0B
19:33:33.082 MBS: MBR Driver receive error 9
19:33:33.380 MBS: Serial Received: 0B 03 00 06 10 00 02 0F 06 1F
19:33:33.382 MBS: MBR Driver receive error 9
19:33:33.430 MBS: Serial Received: 0B 03 04 1E 1A 11 01 05 09 03
19:33:33.431 MBS: MBR Driver receive error 9
19:33:33.593 MBS: Serial Received: 0B 03 00 06 00 02 06 1E 0E 0B 03 04 0D 09 0D 1A 1A 1A 07 0D 01 00 00 00 1E 1A 11 01 19 1A 11 01 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 1E 1A 11 01 19 1A 11 01 01 00 00 00 1F 02 00 00 1F 02 00 00 1F 02 00 00 00 00 00 00 02 1D 0D 1B 1D 0A 13
19:33:33.596 MBS: MBR Driver receive error 9
19:33:33.739 MBS: Serial Received: 0B 03 00 06 10 00 02 0F 06 1F 0B 03 04 1E 1A 11 01 05 09 03
19:33:33.741 MBS: MBR Driver receive error 9

**Reply by danielkucera** [reply]
@AndresL71 , can you check if you have correctly configured `ModbusBaudrate` and `ModbusSerialConfig` ?
For me the serial config was exceptionally off.

**Reply by AndresL71** [reply]
21:27:49.849 CMD: Grp 0, Cmd 'MODBUSSERIALCONFIG', Idx 1, Len 1, Pld 3, Data '3'
21:27:49.857 MBS: Serial UART2
21:27:49.858 MBS: MBR SW ser init at 9600 baud
21:27:49.861 RSL: RESULT = {"ModbusSerialConfig":"8N1"}
21:27:49.961 MBS: Serial Received: 00 00 00 00 00 00 00 00 00 00 00 00 3F 1F B7 E9 3F 1F B7 E9 00 00 00 00 BF 80 00 00 BF 80 00 00 BF 80 00 00 00 00 00 00 42 47 B8 52 77 19 
21:27:49.963 MBS: MBR Driver receive error 9
21:27:50.023 MBS: Serial Received: 0B 03 20 06 00 02 2F 60 0B 03 04 3F 1F C5 05 FE B2 
21:27:50.025 MBS: MBR Driver receive error 9

**Reply by danielkucera** [reply]
This one is for example correct:
```
21:27:50.023 MBS: Serial Received: 0B 03 20 06 00 02 2F 60 0B 03 04 3F 1F C5 05 FE B2
```
`0B 03 20 06 00 02` has crc: `2F 60` 
`0B 03 04 3F 1F C5 05` has crc: `FE B2`
But it is incorrectly parsed as one message.

**Reply by danielkucera** [reply]
This needs to be somehow implemented:
https://electronics.stackexchange.com/questions/371839/detect-modbus-rtu-start-of-frame-and-end-of-frame

**Reply by danielkucera** [reply]
I somehow did.
@AndresL71 , can you pull latest changes and test my branch? But be warned, some devices do not respect this delay as I have found.

**Reply by AndresL71** [reply]
still geting the same result 

08:23:33.511 MBS: Serial Received: 0B 03 20 06 00 02 2F 60 0B 03 04 3E C1 A3 6E F5 3B 
08:23:33.513 MBS: MBR Driver receive error 9
08:23:34.161 MBS: Serial Received: 0B 03 20 06 00 02 2F 60 0B 03 04 3E BF BE 77 5D B9 
08:23:34.163 MBS: MBR Driver receive error 9
08:23:34.461 MBS: Serial Received: 0B 03 20 06 00 02 2F 60 0B 03 04 3E C0 AA 65 E3 6C 
08:23:34.463 MBS: MBR Driver receive error 9
08:23:34.860 MBS: Serial Received: 0B 03 20 06 00 02 2F 60 0B 03 04 3E BE 0D ED F9 22 
08:23:34.862 MBS: MBR Driver receive error 9
08:23:35.486 MBS: Serial Received: 0B 03 20 06 00 02 2F 60 0B 03 04 3E BE 0D ED F9 22 
08:23:35.489 MBS: MBR Driver receive error 9

**Reply by danielkucera** [reply]
@AndresL71 , can you try to experiment with the value 3.5 here?
https://github.com/danielkucera/Tasmota/blob/662a47f383fb71a50cc018ea5bfccaa87fd6080b/lib/lib_basic/TasmotaModbus-3.6.0/src/TasmotaModbus.cpp#L79
Try maybe 2, maybe 1,...

**Reply by AndresL71** [reply]
Different values…. Same results. 
 
 what Hw setup are you using. ?

I am trying to sniff a rs485 channel between a huawei inverter (modus master) and a dsu666 meter (slave)

**Reply by danielkucera** [reply]
Then it probably doesn't comply to modbus specification what seems quite common, see this part:
https://minimalmodbus.readthedocs.io/en/stable/serialcommunication.html#timing-of-the-serial-communications
One of my devices has following timing, there is only 560us between the messages so a passive listener cannot detect where those messages end - this gives crc error:

![RigolDS10](https://github.com/arendst/Tasmota/assets/1734256/b4c0140b-2882-441d-9bc4-043723687e4f)

But another device I have has a generous delay (30ms) between command and response and is easily decoded:

![RigolDS11](https://github.com/arendst/Tasmota/assets/1734256/ada4d746-36a7-477d-a407-eb897006dda0)

I don't know how to handle the first situation. One way is maybe ignore the failed checksum (because when you merge two messages, the checksum at the end obviously doesn't match) and send it as one. Or maybe specify how many bytes to expect based on the header but that excludes custom message decoding and would be very complicated.

**Reply by danielkucera** [reply]
Hi @AndresL71 ,
I tried another approach with checking CRC after each received byte.
Can you try if that works for you?

**Reply by AndresL71** [reply]
similar results

03:54:41.696 MBS: Serial Received: 0B 03 00 06 00 02 06 1E 0E 0B 03 04 0D 05 11 0A 13 00 09 05 0B 02 00 00 00 1F 02 14 07 11 02 14 07 11 00 00 00 1D 0A 09 05 10 0A 0A 09 05 10 02 00 00 00 1F 0A 14 14 05 1F 0A 14 
03:54:41.711 MBS: MBR Driver receive error 9
03:54:41.743 MBS: Serial Received: 14 05 00 00 00 00 1F 1A 11 19 0F 1A 11 19 0F 1A 11 19 0F 00 00 00 02 01 0A 01 1D 0A 06 1F 
03:54:41.746 MBS: MBR Driver receive error 9
03:54:41.843 MBS: Serial Received: 0B 03 00 06 10 00 02 0F 06 1F 
03:54:41.845 MBS: MBR Driver receive error 9
03:54:41.892 MBS: Serial Received: 0B 03 04 1F 02 14 16 1A 13 1F 
03:54:41.895 MBS: MBR Driver receive error 9
03:54:42.143 MBS: Serial Received: 0B 03 00 06 10 00 02 0F 06 1F 
03:54:42.145 MBS: MBR Driver receive error 9
03:54:42.192 MBS: Serial Received: 0B 03 04 1F 02 14 16 1A 13 1F 
03:54:42.194 MBS: MBR Driver receive error 9
03:54:42.443 MBS: Serial Received: 0B 03 00 06 10 00 02 0F 06 1F

**Reply by danielkucera** [reply]
@AndresL71 , do you also see some frames correctly received? 

To me it seems like some electrical issue because in the data I cannot see any message with a correct CRC.
Unlike in the one you sent earlier: https://github.com/arendst/Tasmota/discussions/18618#discussioncomment-8978135
